src/utils/shallowEqual.js

```diff
@@ -23,28 +23,26 @@ function shallowEqual(objA, objB) {
     return true;
   }
 
-  if (!objA || !objB) {
+  if (typeof objA !== 'object' || objA === null ||
+      typeof objB !== 'object' || objB === null) {
     return false;
   }
 
-  if (typeof objA !== 'object' || typeof objB !== 'object') {
+  var keysA = Object.keys(objA);
+  var keysB = Object.keys(objB);
+
+  if (keysA.length !== keysB.length) {
     return false;
   }
 
-  var key;
   // Test for A's keys different from B.
-  for (key in objA) {
-    if (objA.hasOwnProperty(key) &&
-        (!objB.hasOwnProperty(key) || objA[key] !== objB[key])) {
-      return false;
-    }
-  }
-  // Test for B's keys missing from A.
-  for (key in objB) {
-    if (objB.hasOwnProperty(key) && !objA.hasOwnProperty(key)) {
+  var bHasOwnProperty = Object.prototype.hasOwnProperty.bind(objB);
```

Comment:
yeah, .bind sometimes get the problem with GC when you use this functions with DOM listeners
in this case i tested on 120-length hashes, calling 30000+ times
version with .bind showed better results on different tests
maybe you know testes, where it can create a problems?